### PR TITLE
Add genrics store function, modify parse annotation path to search an…

### LINKED ISSUE #26: Make `Store` trait methods generic
`Store` trait has multiple `save` and `load` methods specific to each resource type (e.g. `Pod`, `PodJob`). For a specific implementation, we can simplify the implementation by invoking a generic store/load method that generically works against all of these resource types.

### LINKED ISSUE #27: Make `Annotation` optional
For example, this would make it easier to instantiate a `Pod` and load from store.

Some considerations:
- `load_pod`
  - w/ name/version -> loads annotation + spec into instance
  - w/ hash -> loads spec only into instance
- `save_pod`
  - w/ annotation -> saves both annotation + spec
  - w/o annotation -> saves only spec
- `delete_pod`
  - w/ hash -> delete spec dir + all annotations that refer to it (if annotation dir becomes empty, clean up as well)
  - w/ name/ver -> delete annotation only (if annotation dir becomes empty, clean up as well)
- `list_pod` -> make name and ver `Vec<Option<_>>` to account for `None` on "spec-only" pods

### LINKED ISSUE #36: Change Annotation handling
In the running design, `Annotation` type lives in its own directory, separate from the object it's annotating. This makes moving the target object with annotation difficult. In the new design, `Annotation` will directly live in the resource folder that contains the target resource. For example, an annotation for a `pod` would be found as follows in the filestore:
```
pod
  └──hash
               ├── spec.yaml (Pod resource info)
               └── name-version.yaml  (annotation yaml)

Exact specification of the annotation file name is YTD but multiple annotation files would be allowed.
This issue is related to #27 as a resource may lack any annotation yaml, making annotation optional.

### PR DESCRIPTION
**Pull request by Synicix**
-Add genrics store function
-modify parse annotation path to search annotation
- fix delete pod missing handle case and convert it to use generics
- made get_struct_name panic if fail to extract
- fix regex
- remove regex not found

### CODE REVIEW

**Comment by Synicix** [comment]
#26

**Comment by Synicix** [comment]
@guzman-raphael @eywalker Going to open it up for discussion on either tomorrow or Friday

For the most part it is ready for review and merge, but would like your feedback on how to improve it.

**Comment by Synicix** [comment]
Note #35 will depend on this, since it uses a lot of the generic functions I made

File: src/util.rs
Comment (eywalker) [inline src/util.rs]: Actually why are we getting position 2?

**Review by eywalker — CHANGES_REQUESTED** [review]
I agree with the strict name lookup not returning a result. However I'd want to reassess its use and logic.

File: src/util.rs
Comment (Synicix) [reply src/util.rs]: Okay just checked the code, since I didn't really looked at it before after it got merge into the initial pull request.

So, for example, given a pod, type_name::<Pod>() returns "orcapod::model::Pod", so we would need to extract the last section and ignore the rest. Given that, this should have been implemented with last and not get(2). I will make a change quickly

**Comment by Synicix** [comment]
@eywalker @guzman-raphael 
Summary:
- Change annotation design to save in the item folder
- Add cacheing in form of BTreeMap<name-ver, item_hash> with binary encoding and race condition handling for file system io

TODO:
- Need to add glob and regex to construct BTreeMap in case there is a change done outside the system.
- Error Handling (might leave till later)

I have updated the pull request to support the new design of annotation with a slight modification to make the folder structure cleaner.

What we agree to try yesterday was to save annotation in the item folder, but instead of doing that, I now save it under all annotations under its own folder, under the item folder.

i.e
/pod/annotations/<name>-<ver>.yaml
The reason behind this because I didn't want to gunk up the main folder with a bunch of annotations that user may not care about.

As for caching, I implemented a BTreeMap<name-ver, item_hash> along with storing it in a binary and the associated file locking system to prevent race conditions when more than one instance of the program is updating the localfilestore.

File: .vscode/launch.json
Comment (Synicix) [inline .vscode/launch.json:1]: Revert the change

File: src/store/filestore.rs
Comment (Synicix) [inline src/store/filestore.rs]: Look into join

File: src/store/filestore.rs
Comment (Synicix) [inline src/store/filestore.rs]: rename to make_annotation_path

File: src/store/filestore.rs
Comment (Synicix) [inline src/store/filestore.rs]: change to 
load_item<T:asjdfklajsdf>(&self, spec; ResourceSpec)

enum ResourceSpec {
hash(string,
annotation(String, String)
}

File: src/store/filestore.rs
Comment (Synicix) [inline src/store/filestore.rs]: Match ResourceSpec above

Comment (Synicix) [inline src/store/filestore.rs]: Explicit delete the item regardless if there are annotation pointing to it.

Comment (Synicix) [reply src/store/filestore.rs]: Also add delete annotation function

Comment (Synicix) [reply src/store/filestore.rs]: ^ fine with generic

File: src/store/mod.rs
Comment (Synicix) [inline src/store/mod.rs]: Be removed

**Review by Synicix — COMMENTED** [review]
TODO changes

File: .vscode/launch.json
Comment (Synicix) [reply .vscode/launch.json:1]: Probably need to be updated

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs]: did join

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs]: reanmed

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs]: updated

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs]: Implemented generic

File: src/store/mod.rs
Comment (Synicix) [reply src/store/mod.rs]: removed

**Comment by Synicix** [comment]
Ready for review with the new implementation of cache @eywalker @guzman-raphael

File: .vscode/launch.json
Comment (guzman-raphael) [inline .vscode/launch.json]: You should modify this on your side as needed but we shouldn't change what is committed here since it is an example. Changes should be only if there is a bug or if there is an improvement to the example.

File: Cargo.toml
Comment (guzman-raphael) [inline Cargo.toml]: Rather than adding a dependency for this, you could have added an alias with one line in `error.rs`. Then you could import it everywhere.

For example:

```rust
pub type Result<T> = std::result::Result<T, std::error::Error>;
```

File: Cargo.toml
Comment (guzman-raphael) [inline Cargo.toml]: We shouldn't add an exception for this error. All clippy attributes added should include a reason/justification to document why it was necessary for that case.

File: src/util.rs
Comment (guzman-raphael) [inline src/util.rs]: See comment above. Reason should not have been removed.

File: Cargo.toml
Comment (guzman-raphael) [inline Cargo.toml]: I've just tried removing this on my side and there was no error. Perhaps this change is now stale so we should remove it. Also, when adding a new one, we should use a comment to give context why this was necessary.

File: Cargo.toml
Comment (guzman-raphael) [inline Cargo.toml]: I've just tried removing this on my side and there was no error. Perhaps this change is now stale so we should remove it. Also, when adding a new one, we should use a comment to give context why this was necessary.

File: Cargo.toml
Comment (guzman-raphael) [inline Cargo.toml]: I've just tried removing this on my side and there was no error. Perhaps this change is now stale so we should remove it. Also, when adding a new one, we should use a comment to give context why this was necessary.

File: Cargo.toml
Comment (guzman-raphael) [inline Cargo.toml]: I've just tried removing this on my side and there was no error. Perhaps this change is now stale so we should remove it. Also, when adding a new one, we should use a comment to give context why this was necessary.

File: src/store/mod.rs
Comment (guzman-raphael) [inline src/store/mod.rs]: Why was it necessary to make `self` a `mut`? Because of our immutability design, we've generally found ways of not needing this. I'm missing some context/justification here.

File: src/store/mod.rs
Comment (guzman-raphael) [inline src/store/mod.rs]: Unclear why the output type changed here. Keep in mind the idea was to use this output to display a table to the user (UI or CLI).

A typical programming pattern is to represent tables as as an ordered mapping where keys are the column name and value is an array of row values (all with the same length). This scales well, is easy to modify, and gives greater flexibility to the caller.

I think I remember you saying that the value might have become a concatenated value of 2 fields? If so, not sure I understand why it is better to always require string splitting of each row by the caller if they need to consume them.

File: src/store/mod.rs
Comment (guzman-raphael) [inline src/store/mod.rs]: Assuming `delete_annotation` below is the explicit delete of 1 annotation, would this then be an explicit delete of a whole pod i.e. the pod spec directory + all annotation references?

If so, then it might be better to change the input signature to just require a `hash` argument.

File: src/error.rs
Comment (guzman-raphael) [inline src/error.rs]: We discussed this at some point with @eywalker. It is isn't required by Rust since you can inline import, however, we mentioned following a good practice for ourselves of including everything that is coming externally at the top through `use` statements for clarity.

Unfortunately clippy doesn't have a lint for the ones that are only 1 level deep...

File: src/store/mod.rs
Comment (guzman-raphael) [inline src/store/mod.rs]: This PR seems to also address making `Annotation` optional. If so, I would expect `load_pod` to handle loading with name/version (with annotation) or using hash (w/o annotation). The corresponding [issue](https://github.com/walkerlab/orcapod/issues/27) has more info on these cases.

My recommendation would actually be to move that feature into a separate PR since it is starting to become too much for one PR. That way it isn't all-or-nothing.

**Review by guzman-raphael — CHANGES_REQUESTED** [review]
@Synicix Thanks for the updates. :handshake:

Submitting PR review progress for now but still need to review the following when I get back.

- src/model.rs
- src/filestore.rs
- tests/fixture/mod.rs
- tests/model.rs
- tests/store.rs

File: Cargo.toml
Comment (Synicix) [reply Cargo.toml]: Wouldn't that still require you to import Result from error now, when it is not really tied to that? In short anyhow is just a generic wrapper that allow for more stack tracing that takes any error type.

File: Cargo.toml
Comment (Synicix) [reply Cargo.toml]: I removed it mainly because of test that would require a lot of these exception when it is pretty obvious as to the reason why. Added it back in

File: Cargo.toml
Comment (Synicix) [reply Cargo.toml]: It had to do with fs4 and colorized. I remove fs4 for now since we are not doing file operations, but I can include the reason why

File: Cargo.toml
Comment (Synicix) [reply Cargo.toml]: Yeah this was an issue with fs4, I removed it because it was complaining that
``` rust
let file = fs::OpenOptions::new().read(true).open("foo.txt");
file.exclusive_lock()
```

should be
```
fs::read(<file>)
```

The issue is that, fs4 requires verbose file reads, and we will most likely run into it later when we start to use that. But for not I can remove it

File: Cargo.toml
Comment (Synicix) [reply Cargo.toml]: Odd, before than it was complaining about the regex. Guess when I rebuilt the container it might have became stale with the new standards. Removed.

File: Cargo.toml
Comment (Synicix) [reply Cargo.toml]: Issue with this was missing_panic_doc was triggering on the helper functions in test that do panic due to unwrap()
```
error: docs for function which may panic missing `# Panics` section
  --> tests/fixture/mod.rs:25:5
   |
25 |     pub fn get_name(&self) -> &str {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first possible panic found here
  --> tests/fixture/mod.rs:27:32
   |
27 |             Self::Pod(pod) => &pod.annotation.as_ref().unwrap().name,
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_panics_doc
   = note: `-D clippy::missing-panics-doc` implied by `-D clippy::pedantic`
   = help: to override `-D clippy::pedantic` add `#[allow(clippy::missing_panics_doc)]`

error: docs for function which may panic missing `# Panics` section
  --> tests/fixture/mod.rs:38:5
   |
38 |     pub fn get_version(&self) -> &str {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first possible panic found here
  --> tests/fixture/mod.rs:40:32
   |
40 |             Self::Pod(pod) => &pod.annotation.as_ref().unwrap().version,
   |                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_panics_doc

error: docs for function which may panic missing `# Panics` section
  --> tests/fixture/mod.rs:51:5
   |
51 |     pub fn set_name(&mut self, name: &str) {
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: first possible panic found here
  --> tests/fixture/mod.rs:53:52
   |
53 |             Self::Pod(pod) => name.clone_into(&mut pod.annotation.as_mut().unwrap().name),
   |                                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_panics_doc
```

I added the reason for adding it

File: src/store/mod.rs
Comment (Synicix) [reply src/store/mod.rs]: We have to make it mut, due to self-containing the cache structure which needs to be updated in memory. The previous version stored it on the disk hence why we were able to get away with it.

Comment (Synicix) [reply src/store/mod.rs]: I think discussion can be continue on slack

File: src/store/mod.rs
Comment (Synicix) [reply src/store/mod.rs]: Currently it just forward back the cache which is comprise of BtreeString<(name-ver), (hash)> for easy look up and table construction.

I would be okay with converting where keys become the column name. What do you think @eywalker

File: src/store/mod.rs
Comment (Synicix) [reply src/store/mod.rs]: delete annotation, will only delete annotation and not the underlying object. Going off of what I remember we discussed last week. Delete_pod will wipe the entire item along with all annotations

File: src/error.rs
Comment (Synicix) [reply src/error.rs]: I think the reason why I removed it was because it conflicts with Error thus we need to explicitly define it as serde_yaml::Error making the import moot. I can add it back in for the sake of clairty

File: src/store/mod.rs
Comment (Synicix) [reply src/store/mod.rs]: A few of the changes deals provides the ability to do that a bit easier in the future, but since we already decided on going with enum to identify it, I can code it in by tonight.

Comment (Synicix) [reply src/store/mod.rs]: Update pull request to have handle this case along with tests.

**Comment by Synicix** [comment]
Removed cache for now @guzman-raphael @eywalker  Ready for review

File: src/model.rs
Comment (guzman-raphael) [inline src/model.rs:45]: I see you removed the skip of the first line which had the `class` attribute. Is it not needed anymore? Is `serde_yaml` smart enough to see that in `T` (`Pod`) it isn't defined and therefore it skips it automatically?

File: src/model.rs
Comment (guzman-raphael) [inline src/model.rs]: What was the reason you needed `PartialEq` and `Clone` on several of these? Was it for asserts in tests?

File: src/model.rs
Comment (guzman-raphael) [inline src/model.rs]: Rather than doing this, I think it would be better to make `annotation` optional in `new`'s arguments as well so you can initialize a `pod` w/o annotation too. Then, it would should be a pass-through like the others.

File: src/store/filestore.rs
Comment (guzman-raphael) [inline src/store/filestore.rs]: This still seems like a cache implementation builtin to the logic since it applies a copy.

File: src/store/filestore.rs
Comment (guzman-raphael) [inline src/store/filestore.rs]: `item` might not be clear/intuitive. We've been calling the core classes *models* which was borrowed from [Pydantic](https://docs.pydantic.dev/latest/concepts/models/) community. Perhaps we can rename `item` to `instance` similarly to how they refer to it?

File: src/store/filestore.rs
Comment (guzman-raphael) [inline src/store/filestore.rs]: The intent behind `make_annotation_path`, `make_spec_path`, and `parse_annotation_path` was so that there was no other mention of paths anywhere so that any change to paths or to file naming would just affect those methods so it takes minimal work to make updates later.

It seems odd that we expect the caller to *know* the file is called `spec.yaml` which also breaks in more places if we change this later.

File: src/store/filestore.rs
Comment (guzman-raphael) [inline src/store/filestore.rs]: Similar to my previous feedback [here](https://github.com/walkerlab/orcapod/pull/34#discussion_r1812862407), I still prefer having tabular data represented as a `BTreeMap<String, Vec<_>>` where the key is column name and the vector is the values for the rows.

File: src/store/filestore.rs
Comment (guzman-raphael) [inline src/store/filestore.rs:64]: Why is this better? Is there some documentation you could point me to?

File: Cargo.toml
Comment (guzman-raphael) [reply Cargo.toml]: We shouldn't rely on an external dependency when we can do this ourselves with just one line. Also, this will be changed anyway once #37 is merged.

File: src/store/filestore.rs
Comment (guzman-raphael) [inline src/store/filestore.rs:190]: In general, it is *usually* better to perform an iterator (lazy) approach as opposed to explicit loops. Here are a few reference articles for Rust specifically that can give some context where I am coming from.

- https://www.lurklurk.org/effective-rust/iterators.html
- https://doc.rust-lang.org/book/ch13-04-performance.html

**Review by guzman-raphael — CHANGES_REQUESTED** [review]
Updating my review. Haven't gone through the tests much since I'll probably do a follow-up pass over it after merge anyway to help out.

File: src/model.rs
Comment (Synicix) [reply src/model.rs:45]: Yes, if the field is not defined in the class it skips it and doesn't load it in.

File: Cargo.toml
Comment (Synicix) [reply Cargo.toml]: Perhaps I didn't do a good job of explaining it. The anyhow Error is a wrapper around the std::error that has some additional feature regarding backtrace. It doesn't really change the behavior but instead add on to it.

More info can be found here: https://docs.rs/anyhow/latest/anyhow/

File: src/model.rs
Comment (Synicix) [reply src/model.rs]: Fair, since we are going to use the none annotation later on anyways. Will change

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs]: Not sure instance is a good name for this either since isn' that also pretty generic too?

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs]: One problem with this is the vectors will be locked a specific type such as String. I assume we are okay with that? A vector of struct would be more flexiable in the the future instead of a btreemap

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs:64]: We don't really want to force it to covert into PathBuf, instead we just want it to implment the Path reference.
Pathbuf is like String and path is like &str. We got the idea from how std::path lib implments it: https://doc.rust-lang.org/src/std/path.rs.html#3262

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs:190]: I prefer readability in this case since I didn't want to redefine the regex pattern every loop like you had in your previous example. Plus it had that odd return type.

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs]: The comment is stale, it still uses the same structure as before, just without storing it in memory. The Btreemap is just a nice way to quickly look up annotation-version -> hash. I will update the comment

**Comment by Synicix** [comment]
@eywalker @guzman-raphael Ready for review again, but probably need to discuss on some of the points above.

File: src/error.rs
Comment (eywalker) [inline src/error.rs:14]: Let's discuss the use of `anyhow` crate along with whether to make this alias public.

File: src/error.rs
Comment (eywalker) [inline src/error.rs:63]: Why was this removed?

File: src/store/filestore.rs
Comment (eywalker) [reply src/store/filestore.rs]: we could just use the term `model` no?

File: src/store/filestore.rs
Comment (eywalker) [reply src/store/filestore.rs:64]: Yeah this allows for the method to more broadly accept different input types. It is a bit more idiomatic to use the following: PathBuf::from(directory.as_ref())

File: src/store/filestore.rs
Comment (eywalker) [reply src/store/filestore.rs:190]: Let's put this as a point for a quick discussion

**Review by eywalker — CHANGES_REQUESTED** [review]

File: src/error.rs
Comment (Synicix) [reply src/error.rs:63]: Does regex function I rewrote allow for no mismatch and relies on it. So instead of erroring out it just returns and empty Btree

File: Cargo.toml
Comment (guzman-raphael) [reply Cargo.toml]: @Synicix is looking into why the alias isn't working when set in the library.

File: src/store/filestore.rs
Comment (guzman-raphael) [reply src/store/filestore.rs]: Discussed with @Synicix. Planning to go with `item` -> `model` and `ItemKey` -> `ModelID`.

File: src/store/filestore.rs
Comment (guzman-raphael) [inline src/store/filestore.rs]: @Synicix Is going to change the return type to a `Vec<Struct>` and we'll revisit once the update is in.

File: src/store/filestore.rs
Comment (guzman-raphael) [reply src/store/filestore.rs:190]: @guzman-raphael Will look into a few cases and report back.

File: src/store/filestore.rs
Comment (eywalker) [reply src/store/filestore.rs]: Sounds good!

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs]: Done

File: src/model.rs
Comment (Synicix) [reply src/model.rs]: Yep that is the main reason

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs]: Updated with static const now

File: src/store/filestore.rs
Comment (Synicix) [reply src/store/filestore.rs]: Chhange to vector of struct

**Comment by Synicix** [comment]
@guzman-raphael @eywalker So with the discussion that Rapahel and me had. I resolve some of the change requested with a few remaining.

- Unable to get the anyhow alias to work due to the fact that the return error type must be of anyhow::Error for the stack trace to work. Thus it has to be used on the top level for it to display to force conversion of our OrcaError into anyhow::Error
- Waiting for Raphael to see if there is a better implemnetation with maps regarding the name_ver_tree
- Test still doesn't do the fixture for the pod yet (might be able to pick this up later tonight or tomrrow)

File: src/error.rs
Comment (eywalker) [reply src/error.rs:14]: I'm going to let this pass for now and we can decide to revisit the use of `anyhow` in the future PR. Fortunately, the usage pattern of `Result<T>` won't change.

File: src/store/filestore.rs
Comment (eywalker) [reply src/store/filestore.rs:190]: In the interest of having some closure on this PR that's been up for a while, let's set this as a separate issue and address in the future.

File: src/store/mod.rs
Comment (eywalker) [inline src/store/mod.rs:4]: We should consider making this clonable

File: src/store/mod.rs
Comment (eywalker) [inline src/store/mod.rs:41]: Let's follow up on the specification of list behavior as a dedicated issue.

**Review by eywalker — APPROVED** [review]
There are a few hanging points that I'd want to move as issues to be followed up on so that this PR can be merged. This branch has already been serving as the base for many development and it should be merged.

**Review by guzman-raphael — APPROVED** [review]
Merging for now to unblock and address the remaining points in a follow-up PR.

**Pull request merged**
